In [1]:
import numpy as np
import pandas as pd
import math

In [2]:
data = np.array([2,3,3,4,4,4,4,5,5,6])
data

array([2, 3, 3, 4, 4, 4, 4, 5, 5, 6])

In [12]:
# 함수 이용
a = np.array([np.mean(data), np.var(data,ddof=0), np.std(data), np.max(data), np.min(data), np.percentile(data,25), np.median(data), np.percentile(data,75)]).reshape(1,8)
df_a = pd.DataFrame(a,
                columns=['평균', '불편분산', '표준편차', '최대', '최소', '1분위수', '중앙값','3분위수'],
                index=['해당 값'])
df_a

평균  불편분산      표준편차   최대   최소  1분위수  중앙값  3분위수
해당 값  4.0   1.2  1.095445  6.0  2.0  3.25  4.0  4.75

In [4]:
# percentile 함수의 설명을 보기 위해 버전 확인..
np.__version__

'1.21.2'

- numpy percentile 함수의 인수들:
https://numpy.org/doc/stable/reference/generated/numpy.percentile.html
- 1.20버전:
https://numpy.org/doc/1.20/reference/generated/numpy.percentile.html
- 선형보간법:
https://en.wikipedia.org/wiki/Linear_interpolation
- 정리(i와 j 사이)
    - linear: i + (j - i) * fraction (기본값)
    - lower: i
    - higher: j
    - nearest: i or j whichever is nearest
    - midpoint: (i + j) / 2

In [5]:
# 선형 보간법을 이용한 1분위수 (default)
np.percentile(data,25,interpolation='linear')

3.25

In [6]:
# 최단입점 보간법을 이용한 1분위수
np.percentile(data,25,interpolation='nearest')

3

In [7]:
# 함수 없이 직접 구하기 (인간적으로 sum이나 len 정도는 사용하였습니다...)
import math
b = np.zeros_like(a)
df_b = pd.DataFrame(b,
                columns=['평균', '불편분산', '표준편차', '최대', '최소', '1분위수', '중앙값','3분위수'],
                index=['해당 값'])
mu = np.sum(data)/len(data)
df_b.평균 = np.sum(data)/len(data)
df_b.불편분산 = np.sum((data-mu)**2)/(len(data)-1)
df_b.표준편차 = np.sqrt(np.sum((data-mu)**2)/(len(data)))
df_b.최대 = np.max(data) # 이것도 함수 없이 가능하다. for i in data: max = 0 ; if i > max: max = i 의 형태
df_b.최소 = np.min(data)

def medi(a): # 일반적으로 생각하는 중앙값 공식..
    if len(a) % 2 ==0: return (a[len(a)//2-1]+a[len(a)//2])/2
    else: return (a[len(a)//2])

def perc(arr,q): # 분위수(선형보간법 이용) 계산
    fraction = (len(arr)-1)*(q/100) - math.floor((len(arr)-1)*(q/100))
    i = int((len(arr)-1)*(q/100) - fraction)
    return arr[i] + (arr[i+1]-arr[i])*fraction

df_b['1분위수'] = perc(data,25)
df_b.중앙값 = medi(data)
df_b['3분위수'] = perc(data,75)

df_b

평균      불편분산      표준편차  최대  최소  1분위수  중앙값  3분위수
해당 값  4.0  1.333333  1.095445   6   2  3.25  4.0  4.75

In [8]:
# 추가 - 선형보간법에서의 중앙값?
perc(data,50)

4.0